### DISCUSSION TITLE
sample application installation and plugin problems

### DISCUSSION THREAD

**Discussion by marshmob** [wheels-dev/wheels#1148] [Q&A] [answered]
I'm using a locally installed cf2018 and followed Getting the Example App Up and Running](https://cfwheels.org/blog/getting-the-example-app-up-and-running/) .

I'm getting index error if any plugin is present.  No error if no plugins  and app displays install screen but app then fails trying to find authenticate.

I did need to move the app up one level to the root directory before cfw would run and I'm using MYSQL on a win10 machine

I feel like I'm missing something simple and any help would be appreciated.

Thanks
Lee

-----------------------------------------------------------------------------------------------------------------------------------------
Error message

<html>
<body>

Index 8 out of bounds for length 8
--
 
 
The error occurred in C:/ColdFusion2018/cfusion/wwwroot/wheels/plugins/initialization.cfm: line 42Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/plugins/initialization.cfm: line 94Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/plugins/initialization.cfm: line 26Called from C:/ColdFusion2018/cfusion/wwwroot/root.cfm: line 5Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/global/internal.cfm: line 591Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/global/internal.cfm: line 1011Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/global/internal.cfm: line 999Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/events/onapplicationstart.cfm: line 925Called from C:/ColdFusion2018/cfusion/wwwroot/wheels/events/onapplicationstart.cfm: line 881
40 : 	local.plugins = {}; 41 : 	// Within plugin folders, grab info about each plugin and package up into a struct. 42 : 	for (local.folder in $folders()) { 43 : 		// For *nix, we need a case-sensitive name for the plugin component, so we must reference its CFC file name. 44 : 		local.subfolder = DirectoryList("#local.folder.directory#/#local.folder.name#", false, "query");
 

<br class="Apple-interchange-newline">
</body>
</html>

-------------------------------------------------------------------------------------------------------------------------------------------
Sysinfo (no plugins)

<html>
<body>

System
--
CFML Engine | Adobe ColdFusion2018,0,14,330003
CFWheels Version | 2.3.0
Java Runtime | Java(TM) SE Runtime Environment
Java Version | 11.0.1

<br class="Apple-interchange-newline">
</body>
</html>
<html>
<body>

Database
--
Datasource Name | svgapp1
Migrator Adapter Name | MySQL
Product Name | MySQL
Version | 8.0.20
Driver Name | MySQL Connector/J
Driver Version | mysql-connector-java-8.0.20 (Revision: afc0a13cd3c5a0bf57eaa809ee0ee6df1fd5ac9b)
JDBC Version | 4.2

</body>
</html>
<html>
<body>

Paths
--
Event Path | events
File Path | files
Image Path | images
Javascript Path | javascripts
Model Path | models
Plugin Path | plugins
Plugin Component Path | plugins
Stylesheet Path | stylesheets
View Path | views
Controller Path | controllers
Asset Query String |  
Asset Paths |  

<br class="Apple-interchange-newline">
</body>
</html>

**Comment by bpamiri** [comment, marked as answer]
Lee,

Getting the Example app up and running shouldn't be rocket science but unfortunately when you factor in different OSes, different CF engines, and different databases it gets a little complicated. Tom has a fairly thorough explanation of the manual steps involved in the [wiki](https://github.com/cfwheels/cfwheels-example-app/wiki/Installation). Basically you have to start a CF Server, create a database, create a datasource, run the migrations and then you can log in. Have you followed the manual installation steps?

In contrast to that we've created a mechanism to distribute the Example app using ForgeBox.io and CommandBox that takes a lot of the manual steps out. With CommandBox in place, and the CFWheels CLI tools installed, you simply issue:
```
mkdir myApp
cd myApp
install cfwheels-example-app
```

Let me know if you have followed the manual installation steps in the [wiki](https://github.com/cfwheels/cfwheels-example-app/wiki/Installation) or if it's possible to get CommandBox installed to do a quick test.

**Reply by marshmob** [reply]
I installed cfwheels 2.3.0 **only** via zip download into root directory with no plugins and CFW no app seems OK i.e info panels etc.

I installed FlashMessagesBootstrap plugin via commandbox and CFW immediately generates Index 8 out of bounds for length 8 error (as per above)

So I can duplicate the error without using the sample app

I am going nuts trying to do the simplest thing!

Any help would be appreciated

**Reply by bpamiri** [reply]
So are you trying to specifically run the sample app or are you trying to start a new wheels app? Let me see if I can duplicate the error with that plugin. I would say go talk to the blankety blank that published that plugin, but wait, that's me, I think anyway. Let me see what I can figure out.

**Reply by marshmob** [reply]
I was trying to duplicate the problem in the simplest way.  

I think any plugin causes the problem but i must admit that of the 3 that the sample app needs i choose yours.  One less variable to test.

**Reply by bpamiri** [reply]
Okay, I'm able to reproduce the error in the plugin here. Let me look to see what's going on and hopefully resolve the issue with all of them.

**Reply by bpamiri** [reply]
Well, this one turned into an interesting case. The executive summary is that it's an anomoly with query of queries introduced in [2018.0.12+328566](https://www.forgebox.io/view/adobe/version/2018.0.12+328566) version [2018.0.11+326016](https://www.forgebox.io/view/adobe/version/2018.0.11+326016) doesn't exhibit this behavior. 

But let's get to the juicy details. It's true to say I observe the behavior in 2018.0.12 and above and didn't in 2018.0.11, but the more interesting part is what the behavior is. It turns out if you do a query of queries, and have a sort by clause in it, the column from the sort by clause gets duplicated in final query result. This is what is causing the above error. What's fascinating is that the duplication doesn't happen if the column name in the order by clause is upper cased.

Let me try to illustrate what I've observed:
<img width="1661" alt="Screen Shot 2022-06-19 at 7 45 55 AM" src="https://user-images.githubusercontent.com/180555/174486813-e0ad9031-dc6e-47b1-aea9-10ba0814a674.png">

So in the above image I'm tracking and dumping the query that starts as a directory listing of the plugins directory. You see it has 8 columns going through the various code locations except when it gets to the last one where an additional name column has been added duplicating the NAME column already in the query. Also notice the sql command at the bottom that shows the order by clause and the case of the column name.

<img width="1440" alt="Screen Shot 2022-06-19 at 7 50 43 AM" src="https://user-images.githubusercontent.com/180555/174487030-941659bd-fa7f-4562-b6d9-4755054f16f6.png">

In the second image notice the last query also has 8 columns. The duplicated column has been eliminated. All I did was to uppercase the column name in the order by clause to match the case in the query columns. 

Here are a couple of images of the code running under Lucee.
<img width="1308" alt="Screen Shot 2022-06-19 at 8 02 00 AM" src="https://user-images.githubusercontent.com/180555/174487507-c70af5e3-2796-41f9-a3f5-5af620c47017.png">
<img width="1341" alt="Screen Shot 2022-06-19 at 8 09 09 AM" src="https://user-images.githubusercontent.com/180555/174487795-3a565d83-0e91-4566-be2e-de0d3501fa10.png">

You'll notice first of all that Lucee doesn't upper case the column names on the resulting query from a query of queries. And the case of the column name in the order by clause doesn't add a duplicate column. 

So to work around the issue you can for now go to wheels > plugins > [initialization.cfm](https://github.com/cfwheels/cfwheels/blob/develop/wheels/plugins/initialization.cfm). At the bottom of the page are a couple of functions  [$folders](https://github.com/cfwheels/cfwheels/blob/d112949fa098e823ef751263d9eb130cf468e64d/wheels/plugins/initialization.cfm#L279) and [$files](https://github.com/cfwheels/cfwheels/blob/d112949fa098e823ef751263d9eb130cf468e64d/wheels/plugins/initialization.cfm#L293). Change the case of the name column in the ORDER BY clause on both of those functions. This should get you past the hard crash you are getting while I figure out what the best coarse of action might be.

**Reply by bpamiri** [reply]
It seems this is a known issue introduced in [Adobe CF 2021 Update 2 and  2018 Update 12](https://tracker.adobe.com/#/view/CF-4212383). There is a hot fix attached to that ticket for each flavor of ACF but the issue is each successive update did not incorporate the hot fix so you have to reinstall the hot fix if you update between 12, to 13, to 14. That would be a better approach for this than modifying the `initialization.cfm` script that I suggested above.

**Reply by marshmob** [reply]
Applied the hotfix and plugins don't crash CFW - so this problem is fixed Hooray! and thanks.

But during migration table creation works but table/fields updates fail with  
![migration](https://user-images.githubusercontent.com/106727347/174503790-c1f15251-31e8-4e6d-a4b6-a98ddb3a9ec5.JPG)
Any ideas (hopefully an easier question than the initial one)

p.s. 
CommandBox> install cfwheels-example-app points to CFW 2.2.0 not CFW 2.3.0

**Reply by bpamiri** [reply]
This one is probably because of MySQL 8.0. Our test suite throws some errors with MySQL 8.0. It’s on my list of things to get to and creeping up the list as other things are taken care of.

**Reply by marshmob** [reply]
My to do list just keeps getting bigger!

Thanks for all your help

p.s how do I mark a problem solved or answered

**Reply by bpamiri** [reply]
I marked it as answered for you. It's a button in the UI on the first answer in a thread.
